api/v1/controllers/generators.js

```diff
@@ -228,6 +233,17 @@ module.exports = {
       return instance.setCollectors(collectors);
     }) // need reload instance to attach associations
     .then(() => instance.reload())
+    .then((_retInstance) => {
+      /* No need to assign collector if collectors list provided in request
+      already has current collector */
+      if (Array.isArray(collectors) &&
+       collectors.includes(_retInstance.currentCollector)) {
+        return _retInstance;
+      }
+
+      _retInstance.assignToCollector();
+      return _retInstance.save();
```

Comment:
So you weren't able to make it work without calling save right after?
In that case we should probably change back assignToCollector to also save the changes since everywhere it's being called from is just calling save() right after. The one tricky place is in the beforeUpdate hook where it calls assignToCollector if isActive has changed. We would have to move that to the api layer.

Or we could do all the assignment logic in the db layer instead. I'm starting to think that would be better since calling it from the api isn't as straightforward as I expected it to be. Thoughts?

Reply:
"So you weren't able to make it work without calling save right after?" - No, it would have involved going through updateInstance method (which is called by almost all update api's), which is very clean right now and calls save() at last - my changes would have added various checks in the method and would have made it more complex. So, I ended up not touching that.
I can still try and take that route if we are okay with that. What do you think?

"Or we could do all the assignment logic in the db layer instead" - You mean only this logic when we PUT? We can do that, but I feel its more cleaner in api's for this case because we need this only in PUT method (when currentCollector can be deleted from list of collectors). If we add this logic in db hooks, we will be executing the check for all updates. 

"The one tricky place is in the beforeUpdate hook where it calls assignToCollector if isActive has changed. We would have to move that to the api layer." >
How about adding another parameter 'isSavedAfterAssignment' to assignToCollector?

Reply:
Makes sense, I can see why it's a lot more complicated than I was thinking it would be when I changed assignToCollector to not save. I think it's fine as-is.

Reply:
We would be executing the check for all updates, but we can do the check first-thing so it wouldn't really be any additional work:
if (inst.changed('collectors') && !inst.collectors.includes(inst.currentCollector))

Reply:
Adding a parameter would work, but I think it would overly complicate things. I think it would be easier to just move it.

The reason there's a conflict there is because it needs to be one way if it's done in the controller and another way if done in db hooks, so I think always doing it in the same place is the best solution.